src/ssa/goto_unwinder.cpp

```diff
@@ -6,11 +6,169 @@ Author: František Nečas
 /// \file
 /// SSA Unwinder
 
+#include<stack>
+
+#include <goto-programs/remove_skip.h>
+#include <goto-instrument/unwindset.h>
+#include <goto-instrument/unwind.h>
+
 #include "simplify_ssa.h"
 #include "goto_unwinder.h"
 
+/// Puts all unwindings corresponding to a loop into the loop itself.
+/// This is done by changing the backwards goto target to the first instruction
+/// of the outer-most unwind.
+///
+/// Such transformation is necessary for k-induction and BMC to work properly
+/// and be able to prove correctness.
+///
+/// \note The initial state is saved so that it can be later restored.
+///   This is necessary for goto_unwindt to work correctly.
+void goto_local_unwindert::reconnect_loops()
+{
+  std::stack<goto_programt::targett> unwind_starts;
+  for(auto i_it=goto_function.body.instructions.begin();
+      i_it!=goto_function.body.instructions.end(); i_it++)
+  {
+    if(current_unwinding-i_it->get_code().get_int(unwind_flag)==0)
+    {
+      unwind_starts.push(i_it);
+    }
+    if(i_it->is_backwards_goto() && !unwind_starts.empty())
+    {
+      loop_targets[i_it->location_number]=i_it->get_target();
+      i_it->set_target(unwind_starts.top());
+      unwind_starts.pop();
+    }
+  }
+}
+
+/// Resets loop connections to the initial state which was valid before calling
+/// \ref reconnect_loops.
+/// \pre \ref reconnect_loops has been previously called, resulting in
+///   \ref loop_targets map to be filled with loop configuration.
+void goto_local_unwindert::reset_loop_connections()
+{
+  for(auto &i_it : goto_function.body.instructions)
+    if(i_it.is_backwards_goto())
+      i_it.set_target(loop_targets[i_it.location_number]);
+  loop_targets.clear();
+}
+
+/// Marks the newly produced unwinds using an integer flag to ease
+/// processing later.
+/// \param before_unwind: GOTO iterator pointing before the first new unwind.
+/// \param iteration_points: Iteration points as returned by goto_unwindt,
+///   these mark the end of each newly produced unwind.
+/// \param to_unwind: How many unwindings have been made (=the size of
+///   iteration_points).
+void goto_local_unwindert::mark_unwinds(
+  const goto_programt::targett before_unwind,
+  const std::vector<goto_programt::targett> &iteration_points,
+  unsigned to_unwind)
+{
+  auto i_it=before_unwind;
+  i_it++;
+  // Start of the outermost unwind
+  i_it->code_nonconst().set(unwind_flag, to_unwind);
+  unsigned iteration_point=0;
+  while(iteration_point<iteration_points.size())
+  {
+    if(i_it==iteration_points[iteration_point])
+    {
+      i_it++;
+      i_it->code_nonconst().set(unwind_flag, to_unwind-1-iteration_point);
```

Comment:
I will add a comment explaining more in detail the structure of `iteration_points` after unwinding. You are right that this can be a bit confusing and I had to take a look at CBMC's code again to understand what's going on.

The reasoning behind this is that `iteration_points[i]` points to the last instruction of the unwinding, for `i=0`, this is the outermost unwinding. Let me explain on an example, consider a loop whose body (excluding `GOTO`) ends with a `SKIP`, lets unwind it with `n=2`

```
... <---- this is pointed to by before_unwind
IF ... GOTO ...   # outermost unwinding, marked on line 73 as %2
   BODY
   SKIP <--- pointed to by iteration_points[0]
IF ... GOTO ... # inner unwinding, we need to mark it as %1
   BODY
   SKIP <--- pointed to by iteration_points[1]
ORIGINAL LOOP ... # needs to be kept as %0, i.e. without any flag
```
As you can see in the example, by subtracting the iteration point index, we achieve a descending order of the unwind flags which corresponds to %n semantics in SSA.

Reply:
Yeah, so basically we need to reverse the order of unwind numbers. That information would be sufficient in the comment, I think.